### PR TITLE
Format the spec with prettier

### PR DESCRIPTION
**Pull request by benjie**
_This PR has been work in progress for a while; and a lot has changed. Please see the edit history for previous versions of this comment._

Thanks to advances to `spec-md`, we can now format the GraphQL spec with `prettier`! This means consistent formatting throughout the document, and no need for PR comments about there being too many or two few linebreaks :wink: 

Prettier has been desired in the GraphQL spec for [over 3 years](https://github.com/graphql/graphql-spec/pull/403#issue-165727240) - I'm excited to have worked with @leebyron over the past year to have finally made this a reality! :tada: 

This PR is formed of two commits:

1. The first commit adds the tooling around prettier, including the `yarn format` (or `npm run format`) command which formats the repository.
2. The final commit simply runs `yarn format` and commits the results.

To view the HTML diff:

1. checkout `main`
2. run `yarn build`
3. run `mv public/draft/index.html before.html` (using `mv` will also ensure that the old output cannot be reused)
4. checkout the final commit
5. run `yarn build`
6. run `mv public/draft/index.html after.html`
7. At this point, before.html and after.html differ due to different `data-source` URLs; let's remove these:
8. run `cat before.html | sed 's/ data-source="[^"]*"//g' > before-nosource.html`
9. run `cat after.html | sed 's/ data-source="[^"]*"//g' > after-nosource.html`

You can now see the versions with data-source removed are identical:

```
$ shasum before.html after.html before-nosource.html after-nosource.html
2380c111a29891fa4e8547ef5cee32d429299982  before.html
280252b0fe1aac6065cb280470efb1a096447fa4  after.html
09a7ae84464cc49f6affab9bd544ad4d687837a1  before-nosource.html
09a7ae84464cc49f6affab9bd544ad4d687837a1  after-nosource.html
```

### CODE REVIEW

**Comment by benjie** [comment]
Below is the diff of the output from just this last step (the running of prettier, after all the other changes had already been made); as you can see, the only changes are the addition of a single character of whitespace to the end of the final `<td>` elements; this is because the table syntax has changed from

```md
Literal Value            | Variables               | Coerced Value
------------------------ | ----------------------- | ---------------------------
`{ a: "abc", b: 123 }`   | `{}`                    | `{ a: "abc", b: 123 }`
`{ a: null, b: 123 }`    | `{}`                    | `{ a: null, b: 123 }`
```

to the much prettier and auto-formatted:

```md
| Literal Value            | Variables               | Coerced Value                        |
| ------------------------ | ----------------------- | ------------------------------------ |
| `{ a: "abc", b: 123 }`   | `{}`                    | `{ a: "abc", b: 123 }`               |
| `{ a: null, b: 123 }`    | `{}`                    | `{ a: null, b: 123 }`                |
```

You'll notice that these final columns now have spaces followed by a pipe (`|`) after them, which is what causes the diff below. Other than this, the output is identical :raised_hands: 

```
--- output.original5.html	2020-06-02 14:00:26.915457142 +0100
+++ output.tweaked.html	2020-06-02 15:24:14.562067318 +0100

@@ -3291,41 +3291,41 @@
 <thead><tr>
 <th>Literal Value </th>
 <th>Variables </th>
-<th>Coerced Value</th>
+<th>Coerced Value </th>
 </tr></thead>
 <tbody>
 <tr>
-<td><code>{ a: &quot;abc&quot;, b: 123 }</code> </td><td><code>{}</code> </td><td><code>{ a: &quot;abc&quot;, b: 123 }</code></td></tr>
+<td><code>{ a: &quot;abc&quot;, b: 123 }</code> </td><td><code>{}</code> </td><td><code>{ a: &quot;abc&quot;, b: 123 }</code> </td></tr>
 <tr>
-<td><code>{ a: null, b: 123 }</code> </td><td><code>{}</code> </td><td><code>{ a: null, b: 123 }</code></td></tr>
+<td><code>{ a: null, b: 123 }</code> </td><td><code>{}</code> </td><td><code>{ a: null, b: 123 }</code> </td></tr>
 <tr>
-<td><code>{ b: 123 }</code> </td><td><code>{}</code> </td><td><code>{ b: 123 }</code></td></tr>
+<td><code>{ b: 123 }</code> </td><td><code>{}</code> </td><td><code>{ b: 123 }</code> </td></tr>
 <tr>
-<td><code>{ a: $var, b: 123 }</code> </td><td><code>{ var: null }</code> </td><td><code>{ a: null, b: 123 }</code></td></tr>
+<td><code>{ a: $var, b: 123 }</code> </td><td><code>{ var: null }</code> </td><td><code>{ a: null, b: 123 }</code> </td></tr>
 <tr>
-<td><code>{ a: $var, b: 123 }</code> </td><td><code>{}</code> </td><td><code>{ b: 123 }</code></td></tr>
+<td><code>{ a: $var, b: 123 }</code> </td><td><code>{}</code> </td><td><code>{ b: 123 }</code> </td></tr>
 <tr>
-<td><code>{ b: $var }</code> </td><td><code>{ var: 123 }</code> </td><td><code>{ b: 123 }</code></td></tr>
+<td><code>{ b: $var }</code> </td><td><code>{ var: 123 }</code> </td><td><code>{ b: 123 }</code> </td></tr>
 <tr>
-<td><code>$var</code> </td><td><code>{ var: { b: 123 } }</code> </td><td><code>{ b: 123 }</code></td></tr>
+<td><code>$var</code> </td><td><code>{ var: { b: 123 } }</code> </td><td><code>{ b: 123 }</code> </td></tr>
 <tr>
-<td><code>&quot;abc123&quot;</code> </td><td><code>{}</code> </td><td>Error: Incorrect value</td></tr>
+<td><code>&quot;abc123&quot;</code> </td><td><code>{}</code> </td><td>Error: Incorrect value </td></tr>
 <tr>
-<td><code>$var</code> </td><td><code>{ var: &quot;abc123&quot; } }</code> </td><td>Error: Incorrect value</td></tr>
+<td><code>$var</code> </td><td><code>{ var: &quot;abc123&quot; } }</code> </td><td>Error: Incorrect value </td></tr>
 <tr>
-<td><code>{ a: &quot;abc&quot;, b: &quot;123&quot; }</code> </td><td><code>{}</code> </td><td>Error: Incorrect value for field <var data-name="b">b</var></td></tr>
+<td><code>{ a: &quot;abc&quot;, b: &quot;123&quot; }</code> </td><td><code>{}</code> </td><td>Error: Incorrect value for field <var data-name="b">b</var> </td></tr>
 <tr>
-<td><code>{ a: &quot;abc&quot; }</code> </td><td><code>{}</code> </td><td>Error: Missing required field <var data-name="b">b</var></td></tr>
+<td><code>{ a: &quot;abc&quot; }</code> </td><td><code>{}</code> </td><td>Error: Missing required field <var data-name="b">b</var> </td></tr>
 <tr>
-<td><code>{ b: $var }</code> </td><td><code>{}</code> </td><td>Error: Missing required field <var data-name="b">b</var>.</td></tr>
+<td><code>{ b: $var }</code> </td><td><code>{}</code> </td><td>Error: Missing required field <var data-name="b">b</var>. </td></tr>
 <tr>
-<td><code>$var</code> </td><td><code>{ var: { a: &quot;abc&quot; } }</code> </td><td>Error: Missing required field <var data-name="b">b</var></td></tr>
+<td><code>$var</code> </td><td><code>{ var: { a: &quot;abc&quot; } }</code> </td><td>Error: Missing required field <var data-name="b">b</var> </td></tr>
 <tr>
-<td><code>{ a: &quot;abc&quot;, b: null }</code> </td><td><code>{}</code> </td><td>Error: <var data-name="b">b</var> must be non&#8208;null.</td></tr>
+<td><code>{ a: &quot;abc&quot;, b: null }</code> </td><td><code>{}</code> </td><td>Error: <var data-name="b">b</var> must be non&#8208;null. </td></tr>
 <tr>
-<td><code>{ b: $var }</code> </td><td><code>{ var: null }</code> </td><td>Error: <var data-name="b">b</var> must be non&#8208;null.</td></tr>
+<td><code>{ b: $var }</code> </td><td><code>{ var: null }</code> </td><td>Error: <var data-name="b">b</var> must be non&#8208;null. </td></tr>
 <tr>
-<td><code>{ b: 123, c: &quot;xyz&quot; }</code> </td><td><code>{}</code> </td><td>Error: Unexpected field <var data-name="c">c</var></td></tr>
+<td><code>{ b: 123, c: &quot;xyz&quot; }</code> </td><td><code>{}</code> </td><td>Error: Unexpected field <var data-name="c">c</var> </td></tr>
 </tbody>
 </table>
 </section>
@@ -3381,25 +3381,25 @@
 <thead><tr>
 <th>Expected Type </th>
 <th>Provided Value </th>
-<th>Coerced Value</th>
+<th>Coerced Value </th>
 </tr></thead>
 <tbody>
 <tr>
-<td><code>[Int]</code> </td><td><code>[1, 2, 3]</code> </td><td><code>[1, 2, 3]</code></td></tr>
+<td><code>[Int]</code> </td><td><code>[1, 2, 3]</code> </td><td><code>[1, 2, 3]</code> </td></tr>
 <tr>
-<td><code>[Int]</code> </td><td><code>[1, &quot;b&quot;, true]</code> </td><td>Error: Incorrect item value</td></tr>
+<td><code>[Int]</code> </td><td><code>[1, &quot;b&quot;, true]</code> </td><td>Error: Incorrect item value </td></tr>
 <tr>
-<td><code>[Int]</code> </td><td><code>1</code> </td><td><code>[1]</code></td></tr>
+<td><code>[Int]</code> </td><td><code>1</code> </td><td><code>[1]</code> </td></tr>
 <tr>
-<td><code>[Int]</code> </td><td><code>null</code> </td><td><code>null</code></td></tr>
+<td><code>[Int]</code> </td><td><code>null</code> </td><td><code>null</code> </td></tr>
 <tr>
-<td><code>[[Int]]</code> </td><td><code>[[1], [2, 3]]</code> </td><td><code>[[1], [2, 3]]</code></td></tr>
+<td><code>[[Int]]</code> </td><td><code>[[1], [2, 3]]</code> </td><td><code>[[1], [2, 3]]</code> </td></tr>
 <tr>
-<td><code>[[Int]]</code> </td><td><code>[1, 2, 3]</code> </td><td>Error: Incorrect item value</td></tr>
+<td><code>[[Int]]</code> </td><td><code>[1, 2, 3]</code> </td><td>Error: Incorrect item value </td></tr>
 <tr>
-<td><code>[[Int]]</code> </td><td><code>1</code> </td><td><code>[[1]]</code></td></tr>
+<td><code>[[Int]]</code> </td><td><code>1</code> </td><td><code>[[1]]</code> </td></tr>
 <tr>
-<td><code>[[Int]]</code> </td><td><code>null</code> </td><td><code>null</code></td></tr>
+<td><code>[[Int]]</code> </td><td><code>null</code> </td><td><code>null</code> </td></tr>
 </tbody>
 </table>
 </section>
@@ -3457,41 +3457,41 @@
 <thead><tr>
 <th>Expected Type </th>
 <th>Internal Value </th>
-<th>Coerced Result</th>
+<th>Coerced Result </th>
 </tr></thead>
 <tbody>
 <tr>
-<td><code>[Int]</code> </td><td><code>[1, 2, 3]</code> </td><td><code>[1, 2, 3]</code></td></tr>
+<td><code>[Int]</code> </td><td><code>[1, 2, 3]</code> </td><td><code>[1, 2, 3]</code> </td></tr>
 <tr>
-<td><code>[Int]</code> </td><td><code>null</code> </td><td><code>null</code></td></tr>
+<td><code>[Int]</code> </td><td><code>null</code> </td><td><code>null</code> </td></tr>
 <tr>
-<td><code>[Int]</code> </td><td><code>[1, 2, null]</code> </td><td><code>[1, 2, null]</code></td></tr>
+<td><code>[Int]</code> </td><td><code>[1, 2, null]</code> </td><td><code>[1, 2, null]</code> </td></tr>
 <tr>
-<td><code>[Int]</code> </td><td><code>[1, 2, Error]</code> </td><td><code>[1, 2, null]</code> (With logged error)</td></tr>
+<td><code>[Int]</code> </td><td><code>[1, 2, Error]</code> </td><td><code>[1, 2, null]</code> (With logged error) </td></tr>
 <tr>
-<td><code>[Int]!</code> </td><td><code>[1, 2, 3]</code> </td><td><code>[1, 2, 3]</code></td></tr>
+<td><code>[Int]!</code> </td><td><code>[1, 2, 3]</code> </td><td><code>[1, 2, 3]</code> </td></tr>
 <tr>
-<td><code>[Int]!</code> </td><td><code>null</code> </td><td>Error: Value cannot be null</td></tr>
+<td><code>[Int]!</code> </td><td><code>null</code> </td><td>Error: Value cannot be null </td></tr>
 <tr>
-<td><code>[Int]!</code> </td><td><code>[1, 2, null]</code> </td><td><code>[1, 2, null]</code></td></tr>
+<td><code>[Int]!</code> </td><td><code>[1, 2, null]</code> </td><td><code>[1, 2, null]</code> </td></tr>
 <tr>
-<td><code>[Int]!</code> </td><td><code>[1, 2, Error]</code> </td><td><code>[1, 2, null]</code> (With logged error)</td></tr>
+<td><code>[Int]!</code> </td><td><code>[1, 2, Error]</code> </td><td><code>[1, 2, null]</code> (With logged error) </td></tr>
 <tr>
-<td><code>[Int!]</code> </td><td><code>[1, 2, 3]</code> </td><td><code>[1, 2, 3]</code></td></tr>
+<td><code>[Int!]</code> </td><td><code>[1, 2, 3]</code> </td><td><code>[1, 2, 3]</code> </td></tr>
 <tr>
-<td><code>[Int!]</code> </td><td><code>null</code> </td><td><code>null</code></td></tr>
+<td><code>[Int!]</code> </td><td><code>null</code> </td><td><code>null</code> </td></tr>
 <tr>
-<td><code>[Int!]</code> </td><td><code>[1, 2, null]</code> </td><td><code>null</code> (With logged coercion error)</td></tr>
+<td><code>[Int!]</code> </td><td><code>[1, 2, null]</code> </td><td><code>null</code> (With logged coercion error) </td></tr>
 <tr>
-<td><code>[Int!]</code> </td><td><code>[1, 2, Error]</code> </td><td><code>null</code> (With logged error)</td></tr>
+<td><code>[Int!]</code> </td><td><code>[1, 2, Error]</code> </td><td><code>null</code> (With logged error) </td></tr>
 <tr>
-<td><code>[Int!]!</code> </td><td><code>[1, 2, 3]</code> </td><td><code>[1, 2, 3]</code></td></tr>
+<td><code>[Int!]!</code> </td><td><code>[1, 2, 3]</code> </td><td><code>[1, 2, 3]</code> </td></tr>
 <tr>
-<td><code>[Int!]!</code> </td><td><code>null</code> </td><td>Error: Value cannot be null</td></tr>
+<td><code>[Int!]!</code> </td><td><code>null</code> </td><td>Error: Value cannot be null </td></tr>
 <tr>
-<td><code>[Int!]!</code> </td><td><code>[1, 2, null]</code> </td><td>Error: Item cannot be null</td></tr>
+<td><code>[Int!]!</code> </td><td><code>[1, 2, null]</code> </td><td>Error: Item cannot be null </td></tr>
 <tr>
-<td><code>[Int!]!</code> </td><td><code>[1, 2, Error]</code> </td><td>Error: Error occurred in item</td></tr>
+<td><code>[Int!]!</code> </td><td><code>[1, 2, Error]</code> </td><td>Error: Error occurred in item </td></tr>
 </tbody>
 </table>
 </section>
```

File: rfcs/InputUnion.md [hunk 1]
```diff
@@ -142,154 +181,183 @@ mutation {
...
+| Arvo             | [Union](https://avro.apache.org/docs/current/spec.html#Unions)                                              | ✅   | ✅    |
+| OpenAPI 3        | [oneOf](https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/)                           | ✅   | ✅    |
+| JSON Schema      | [oneOf](https://json-schema.org/understanding-json-schema/reference/combining.html#oneof)                   | ✅   | ✅    |
+| Typescript       | [Union](http://www.typescriptlang.org/docs/handbook/advanced-types.html#union-types)                        | ✅   | ✅    |
+| Typescript       | [Discriminated Union](http://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions) | ✅   | ✅    |
+| Rust             | [Enum](https://doc.rust-lang.org/rust-by-example/custom_types/enum.html)                                    | ✅   | ✅    |
+| Swift            | [Enumeration](https://docs.swift.org/swift-book/LanguageGuide/Enumerations.html)                            | ✅   | ✅    |
+| Haskell          | [Algebraic data types](http://learnyouahaskell.com/making-our-own-types-and-typeclasses)                    | ✅   | ✅    |
```
Comment (benjie) [inline rfcs/InputUnion.md:218]: This is such an improvement!

**Review by benjie — COMMENTED** [review]
I've now reviewed all the non-spec markdown changes (spec markdown changes were validated by asserting that there is no diff to the generated HTML) and am happy that the changes are fine :+1:

**Comment by benjie** [comment]
I recommend that #832 is merged separately from this PR to make reviewing the git history more straightforward.

**Comment by leebyron** [comment]
@benjie should we merge this one after going through the rest of the pre-cut editorial process? I'm marking PRs I want to place attention on as clear candidates for the next spec cut and anticipate prettier editing merge conflicts.

Inevitably there will be open PRs that will need to rebase and run prettier, but maybe we should try to limit most of those for after the spec cut?

**Comment by benjie** [comment]
@leebyron Yes, lets do this as the very last step before the spec cut :+1: Given the proof that there's no diff _currently_ I think we can expect that changes will be minimal (code formatting, for example). I'm happy to rebase/fix/etc this PR when the time comes (or you can; the final commit is auto-generated so just don't `pick` that one, instead generate it again manually).

**Comment by leebyron** [comment]
We might want to keep an eye on https://github.com/prettier/prettier/issues/10655 being released

**Comment by benjie** [comment]
I've rebased this, upgraded to the latest prettier, and reapplied the changes. Asserting there's no diff is a little more complex now because of the new `data-source` that `spec-md` is adding; however:

```
git checkout main
npm install
npm run build && cp out/index.html before-prettier.index.html
git checkout prettier
npm install
npm run build && cp out/index.html after-prettier.index.html
cat before-prettier.index.html | sed 's/ data-source="[^"]*"//g' > before-prettier-nosource.index.html 
cat after-prettier.index.html | sed 's/ data-source="[^"]*"//g' > after-prettier-nosource.index.html
shasum before-prettier.index.html after-prettier.index.html before-prettier-nosource.index.html after-prettier-nosource.index.html
```

shows that, ignoring `data-source` the generated HTML is identical:

```
c0612337b3750da593fcd050770c1d492d964434  before-prettier.index.html
7d3b2d968f2e2046323a6fbeec39e223f29fbd96  after-prettier.index.html
466a3c4940d848fa2ef5cf432d602f6e7acebf8c  before-prettier-nosource.index.html
466a3c4940d848fa2ef5cf432d602f6e7acebf8c  after-prettier-nosource.index.html
```

I believe this is safe to merge now :+1:

**Comment by benjie** [comment]
This has been updated again and is ready to merge :+1:

**Pull request merged**